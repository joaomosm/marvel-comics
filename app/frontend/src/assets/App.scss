body {
  width: 100%;
  height: 100%;
  background-color: #333333;
}

.navbar {
  position: fixed;
  top: 0;
  width: 100%;
  z-index: 100;
  height: 100px;
  background-color: rgba(240, 30, 15, 0.85);
  text-align: center;

  .navbar-search {
    font-family: 'Roboto', sans-serif;
    color: #888;
    font-size: 12px;
    margin: 0 auto;
    padding: 8px;
    border-radius: 0.2rem;
    background-color: rgb(255, 255, 255);
    border: none;
    width: 300px;
    display: block;
    font-style: italic;
  }
}

.container {
  margin: 120px 60px 60px 60px
}

.footer {
  margin: 60px;

  button {
    background-color: white;
    min-height: 40px;
    min-width: 120px;
    border: 1px solid white;
    border-radius: 2px;
    color: red;
    cursor: pointer;
    font-family: 'Trebuchet MS', sans-serif;
    text-transform: uppercase;
    font-weight: bold;

    &:first-child {
      color: red;
    }
    &:last-child {
      float: right;
    }

  }

  button:disabled,
  button[disabled] {
    opacity: 0.6;
  }
}

.grid-container {
  display: grid;
  grid-template-columns: repeat(auto-fill, 152px);
  grid-gap: 20px;
  align-items: stretch;
}

.grid-container img {
  border: 1px solid #ccc;
  box-shadow: 2px 2px 6px 0px rgba(0, 0, 0, 0.3);
  max-width: 100%;
}

.image-container {
  position: relative;
}

.image {
  display: block;
  height: auto;
}

.image-overlay {
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  height: 207px;
  width: 132px;
  opacity: 0;
  transition: 0.5s ease;
  border: 10px solid red;
  background-color: rgba(30, 30, 30, 0.65);
}

.image-container:hover .image-overlay {
  opacity: 1;
  cursor: pointer;
}

.image-text {
  color: black;
  font-size: 12px;
  position: absolute;
  top: 90%;
  left: 50%;
  width: 100%;
  height: 20%;
  background-color: white;
  -webkit-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  text-align: center;
  font-family: 'Trebuchet MS', sans-serif;
  text-transform: uppercase;
  font-weight: bold
}

.image-svg {
  color: black;
  font-size: 16px;
  position: absolute;
  top: 10%;
  left: 85%;
  width: 30px;
  height: 27px;
  background: url('./heart_off.png');
  -webkit-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  text-align: center;
  
  &:hover{
    background: url('./heart_hover.png');
  }
  
  &:active {
    background: url('./heart_on.png');
    width: 40px;
    height: 37px;
  }

  img.heart {
    border-radius: 0px;
    border: 0px;
  }
}

.loading-container {
  text-align: center;
  
  .loading-spinner {
    position: fixed;
    top: 47%;
    right: 47%;
    height: 80px;
    pointer-events: none;
  }
}

@media (prefers-reduced-motion: no-preference) {
  .loading-spinner {
    animation: loading-spin infinite 3s linear;
  }
}

@keyframes loading-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
