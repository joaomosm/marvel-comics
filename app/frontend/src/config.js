export const API_HOST = 'http://localhost:3000';

export const COMICS_PER_PAGE = 20

export const SESSION_COOKIE_NAME = '_marvel_comics_session'

export const defaultParams = {
  limit: COMICS_PER_PAGE,
  offset: 0,
  order: '-focDate'
}

export const characterDefaultParams = {
  limit: COMICS_PER_PAGE,
  offset: 0,
  order: '-modified'
}