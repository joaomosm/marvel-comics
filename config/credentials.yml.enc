To+elswkYhTtyuYDnIibcH4gx4hDs316xNVrHaYrkJdXw4GxFtT8cXtEupqcAsSraEXZcprTFmPR1UmDTrchktyS0W9kjxDTARxFB6wKcszNNIVBXVES1cb63BA/DLvJ984aubI5disyygt8IAzQia+ajOIAF+K7sRM5Gw3y86u8ohVY6IL++4GrXTkbmMqlPxRAoySyqswZOtlCaTaabv1r9uKp5yGd+GVJOAh1LcpqpF2PJLackmgrOSzSd66j6GP68qtfbwX1SA/PAd90HXdjizOzwiGP/XRXdAwdtsmXwr+1Xvd+s2FcROslXe3W9MurVNvW4QLZFuIDoE3TZpqPA8EY178qvOLz7rqSTbHiUkEtx532y3FznPZt55zY2ZLH+GTeMI33oiw4BltI4+uesurkb0MU187e--JH0D5qchXlZb2OtR--2U1qmik4mx8P5eXZHu5ObQ==