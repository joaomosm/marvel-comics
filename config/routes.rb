Rails.application.routes.draw do
  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html

  resources :comics, only: [:index] do
    get :get_by_character, on: :collection
  end
end
