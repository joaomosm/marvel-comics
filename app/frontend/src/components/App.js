import React, { useEffect, useState, useCallback } from 'react';
import axios from 'axios';
import { debounce } from 'lodash';
import { useCookies } from 'react-cookie';
import moment from 'moment';

import './../assets/App.scss';
import Dashboard from './Dashboard';
import {
  API_HOST,
  COMICS_PER_PAGE,
  SESSION_COOKIE_NAME,
  defaultParams,
  characterDefaultParams,
} from './../config';
import { camelize } from './../utils';
import CookieStatement from './CookieStatement';

const App = () => {
  // --- initializers
  const [isLoading, setIsLoading] = useState(false);
  const [comics, setComics] = useState(null);
  const [metadata, setMetadata] = useState(null);
  const [params, setParams] = useState(defaultParams);
  const [favoriteComics, setFavoriteComics] = useState([]);
  const [cookies, setCookie, _removeCookie] = useCookies([SESSION_COOKIE_NAME]);
  const [error, setError] = useState(null);

  // --- fetch data ---
  const fetchComics = () => {
    setIsLoading(true);
    axios({
      method: 'get',
      url: API_HOST + '/comics',
      params,
    })
      .then((response) => {
        setComics(response.data.comics);
        setMetadata(response.data.metadata);
        setIsLoading(false);
      })
      .catch((error) => {
        handleError(error);
      });
  };

  const fetchFavoriteComics = () => {
    setIsLoading(true);
    axios({
      method: 'get',
      url: API_HOST + '/favorite_comics',
    })
      .then((response) => {
        setFavoriteComics(camelize(response.data));
        setIsLoading(false);
      })
      .catch((error) => {
        handleError(error);
      });
  };

  const fetchByCharacter = () => {
    setIsLoading(true);
    axios({
      method: 'get',
      url: API_HOST + '/comics/get_by_character',
      params,
    })
      .then((response) => {
        setComics(response.data.comics);
        setMetadata(response.data.metadata);
        setIsLoading(false);
      })
      .catch((error) => {
        handleError(error);
      });
  };

  // --- hooks ---
  useEffect(() => {
    !!params.name ? fetchByCharacter() : fetchComics();
  }, [params]);

  useEffect(() => {
    fetchFavoriteComics();
  }, []);

  const debounceQuery = useCallback(
    debounce((text) => {
      if (!!text) {
        setParams({ ...characterDefaultParams, name: text });
      } else {
        setParams(defaultParams);
      }
    }, 1000),
    [],
  );

  const handleChange = ({ target: { value } }) => {
    debounceQuery(value);
  };

  // --- favorites ---
  const filterFavorites = (favoriteComics) => (comicId) =>
    favoriteComics.filter((fc) => Number(fc) === Number(comicId)).length;
  const checkFavorite = filterFavorites(favoriteComics);

  const toggleFavorite = (comic) =>
    checkFavorite(comic.id) ? removeFavorite(comic) : addFavorite(comic);

  const addFavorite = ({ id, title }) => {
    axios({
      method: 'post',
      url: API_HOST + '/favorite_comics/add_favorite',
      params: { comic_id: id, title },
    }).then((response) => {
      fetchFavoriteComics();
    });
  };

  const removeFavorite = ({ id }) => {
    axios({
      method: 'post',
      url: API_HOST + '/favorite_comics/remove_favorite',
      params: { comic_id: id },
    }).then((response) => {
      fetchFavoriteComics();
    });
  };

  // --- pagination ---
  const handleNextPage = () => setParams({ ...params, offset: params.offset + COMICS_PER_PAGE });

  const handlePreviousPage = () =>
    setParams({ ...params, offset: Math.max(params.offset - COMICS_PER_PAGE, 0) });

  // --- error ---
  const handleError = (error) => {
    if (error.response.data === 'Rate Limit Exceeded') {
      setError('rate_limit');
    }
  };

  // --- cookie ---
  const acceptCookie = () =>
    setCookie(SESSION_COOKIE_NAME, true, { expires: moment().add(30, 'minutes').toDate() });

  if(cookies[SESSION_COOKIE_NAME] !== 'true') {
    return <CookieStatement acceptCookie={acceptCookie} cookies={cookies}/>
  }

  return (
    <Dashboard
      isLoading={isLoading}
      handleChange={handleChange}
      comics={comics}
      metadata={metadata}
      favoriteComics={favoriteComics}
      toggleFavorite={toggleFavorite}
      handlePreviousPage={handlePreviousPage}
      handleNextPage={handleNextPage}
      checkFavorite={checkFavorite}
      error={error}
    />
  );
};



export default App;
