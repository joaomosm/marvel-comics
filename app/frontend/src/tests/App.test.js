import React from 'react';
import { render } from '@testing-library/react';
import { Cookies } from 'react-cookie'

import App from './../components/App';
import { SESSION_COOKIE_NAME } from './../config';

const renderComponent = () => render(<App />);

it('matches snapshot', () => {
  const { container } = renderComponent();
  expect(container).toMatchSnapshot();
});

describe('when cookie does not exist', () => {
  it('renders CookieStatement component', () => {
    const { getByTestId } = renderComponent();
    expect(getByTestId('cookie-statement')).toBeVisible();
  });
});

describe('when cookie exists exist', () => {
  beforeEach(() => {
    const cookies = new Cookies();
    cookies.set(SESSION_COOKIE_NAME, 'true');
  })

  it('renders Dashboard component', () => {
    const { getByTestId } = renderComponent();
    expect(getByTestId('dashboard')).toBeVisible();
  });
});
