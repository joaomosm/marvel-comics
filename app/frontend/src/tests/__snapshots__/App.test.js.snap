// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="cookie-statement"
    data-testid="cookie-statement"
  >
    <div
      class="cookie-statement-text"
    >
      Are you ready to take a 
      <i>
        Great Power
      </i>
      ? It comes with 
      <i>
        Great Responsibility
      </i>
      !
      <button
        class="cookie-statement-button"
      >
        Accept
      </button>
    </div>
  </div>
</div>
`;
